Pr10 type slot required migration

brynnz22 opened the pull request:
This PR addresses making the `type` slot required on every class instance: https://github.com/microbiomedata/berkeley-schema-fy24/pull/10

This PR is structured to occur BEFORE the class names change. E.g. the function calls the `omics_processing_set` to give it a `type` `nmdc:DataGeneration`. 

Because some classes are used inlined, a dictionary was created for the slot that uses the inlined class as the range with the key as the slot and the value the class uri. Depending on when this migration takes place and if new data has been added that uses inlined slots that have not been used before, we may need to update the inlined dictionary with any new slots (this already happened to me while working on the code).


Also, because of the collection name discrepancies in mongo vs. the schema, running the `make make-rdf` command will not work on collections with differing names. I had to put a block of code in the `mongo_dump_api_emph.py` script after line 293 that looks like below to look for validation errors (sub whichever collection with a name discrepancy): 
```

if available_vs_selected_collections['selected only']:
                if 'omics_processing_set' in selected_collections:
                    available_selected_collections.append('omics_processing_set')
```

We might want to consider running this after the name change, so testing using `make-rdf` is more straightforward. If we do want to run this after the collection name change migration, we need to change the names of the collections called in the code. Which should be relatively simple.

brynnz22 commented:
@turbomam If I run the make-rdf on just the biosample_set I am still getting that odd error about `chem_administration` not being an array (though it looks like an array to me). I'm not sure what is causing this.

turbomam commented:
Let's try retrieving some data with @mbthornton-lbl's `get-study-related-records` script and running that through the migrator. I can do that in about 90 minutes.

eecavanna commented:
Hi @brynnz22, thanks for implementing the `SchemaView` approach @turbomam suggested in our meeting earlier today. 

Like you mentioned to me this afternoon, this `SchemaView` approach takes a few seconds to run (_per document_, since everything is repeated for each document). The current migration framework requires everything to run for each document, so I don't see a way around this (e.g. I don't see a way we can create the `SchemaView` one time, up front, while still remaining within the current migration framework).

With that in mind, I'd prefer that you wait for the upcoming changes to the migration framework, before finalizing this migrator. Those changes are in an unmerged (unfinished) PR in `nmdc-schema`. Once they are in `nmdc-schema`, I think they'll eventually also be merged (across repos) into this fork.

turbomam commented:
> Those changes _will_ allow for the `SchemaView` to be generated one time

I have prototyped this in 
- https://github.com/microbiomedata/berkeley-schema-fy24/pull/69/files
but will defer to both of you regarding the best implementation for release

eecavanna commented:
> I have prototyped this in...

Thanks, @turbomam! I forgot all about the "constructor" (i.e. what I've been referring to as the constructor; but which I think is technically the "initializer" in Python; i.e. the `__init__` method). Putting "run this once, before doing the agenda" code in there _would_ work.

So, this statement I made yesterday is incorrect:
> The current migration framework requires everything to run for each document. It doesn't allow us to create the SchemaView one time, up front, and reuse it for all documents.

brynnz22 commented:
@eecavanna and @turbomam I updated this using schemaView and Mark's code from the work he did in the nmdc_schema repo to get inferred inlined classes. I still had to do a tiny bit of hardcoding for types that have a `prov` prefix instead of `nmdc` (there was only one slot for this). 

Also, I updated the collections so they align with the NEW collection names. So this migration should now come after the migration that we do to change the collection names.

turbomam commented on nmdc_schema/migrators/migrator_from_X_to_PR10.py:
```diff
@@ -0,0 +1,183 @@
+from nmdc_schema.migrators.migrator_base import MigratorBase
+from linkml_runtime import SchemaView
+
```
old comment. you may have already addressed this.

brynnz22 replied:
@turbomam I had this originally because you mentioned this, but this was giving me an error based on the a yaml file not existing form the import. Once I removed it, everything worked as it should.

See this commit: https://github.com/microbiomedata/berkeley-schema-fy24/pull/66/commits/443f453d7c9a41bf97dfb342d0879b82923b4fcd

turbomam replied:
OK. please share the error message you were getting.

turbomam commented on nmdc_schema/migrators/migrator_from_X_to_PR10.py:
```diff
@@ -0,0 +1,183 @@
+            value_class_uris[value_class] = uri
+
+        # create dictionary of slots and their range's associated class_uris
+        slots_with_inlined_classes = {}
```
old comment. you may have already addressed this.

brynnz22 replied:
Yep, I believe I did this.

brynnz22 commented:
@turbomam yes sounds good. I could not think of any other way of handling that, so I'd love to hear your thoughts about SchemaView()

aclum commented:
@brynnz22  @eecavanna  have you tried to run this migration locally? we ran into some instances where the existing value for a type slot had a typo in the class name. If fixed the ones we found but am not sure if there are others and wanted to make sure that use case was covered by the migrator.

eecavanna commented:
I haven't run any of the Berkeley schema migrators locally yet. Thanks for the heads up.

brynnz22 commented on nmdc_schema/migrators/migrator_from_X_to_PR10.py:
```diff
@@ -0,0 +1,113 @@
+    def add_type_slot_with_class_uri(self, document: dict, class_uri: str, inlined_slots=None):
+        r"""
+            Adds a type slot to each collection with the appropriate class uri as the value. E.g. type: nmdc:Biosample. If a type
+            slot exists, it will overwrite to the types listed below. If the optional dictionary argument inlined_slots is given, then
```
@aclum I documented that if a type slot exists, it will overwrite to the correct types.

brynnz22 replied:
I will look at this deeper to make sure it does do that so any typos will be overwritten.
